set normed vectors to None when model is trained

timbicker opened the pull request:
Fixes #2260 

allows `most_similar()` to be called based on the newly trained vectors during multiple training iterations.

gojomo commented:
Makes sense to me. For completeness, the same thing should happen on Word2Vec & related `train()`s. 

To be an ideal PR, there'd also be a test that fails before the fix(es) are applied. 

I could also imagine people who've done an explicit `init_sims(replace=True)`, then trying to re-train, getting odd or bad results from that training (since they've discarded the original raw magnitudes). This sort of re-set could also catch such a condition – by testing if the normed and non-normed properties are identical objects – and issue a (non-failing because it still might kinda work) warning in such a case.

timbicker commented:
@gojomo thank you for your reply. I added two tests, one each for doc2vec and word2vec. I also wanted to add a test for fast text, but it turns out that there is some error when the model is trained more than once. There is also an issue #2139 for this. Should I still implement the test?

Further, word2vec overrides `_clear_post_train` and sets `self.wv.vectors_norm = None`. So the functionality is already included here. I therefore structured the resetting of the vectors in doc2vec analogously.

menshikh-iv commented on gensim/models/doc2vec.py:
```diff
@@ -649,6 +649,7 @@ def clear_sims(self):
         """Resets the current word vectors. """
         self.wv.vectors_norm = None
         self.wv.vectors_docs_norm = None
+        self.docvecs.vectors_docs_norm = None
```
can you please check and verify that "similar" thing happens in w2v\fasttext models too?

timbicker replied:
Yes. If you look in `word2vec.py` you see that there are actually two functions `_clear_post_train` and `clear_sims` which do the same thing (one of the functions is therefore redundant): Setting `self.wv.vectors_norm = None`.
The same thing (setting `self.wv.vectors_norm = None` and more) also happens in `fasttext.py` in function `_clear_post_train`.
Does this solve your concerns or do you want a more detailed explanation?

timbicker replied:
@menshikh-iv I am not sure, whether I should resolve the conversation or wait for a response by you. What is preferred here (or in general in such cases)?

gojomo replied:
Separately, the line above this (which existed prior to your edits) – `self.wv.vectors_docs_norm = None` – looks fishy. A normal `wv` set of word-vectors wouldn't be expected to even have a `vectors_docs_norm` property, unless some other code has mistakenly poked doc-vectors into there.

menshikh-iv commented on gensim/test/test_doc2vec.py:
```diff
@@ -687,6 +687,22 @@ def testTrainWarning(self, l):
+        model.build_vocab(corpus)
+
+        model.train(corpus, total_examples=model.corpus_count, epochs=1, start_alpha=0.1, end_alpha=0.1)
+        model.docvecs.most_similar(1)
```
it's better to test in the same manner (i.e. using callbacks) as mentioned in https://github.com/RaRe-Technologies/gensim/issues/2260#issue-377788511 (i.e. call most_similar "during training", not after). Same things for Word2Vec and FastText

timbicker replied:
I see your point. The problem is that above mentioned function `_clear_post_train`, which is responsible to reset the previously trained docvecs, is only called after the whole training is done. This means that calls of `most_similar` inside a callback, won't be able to access the newly trained vectors.
One solution would be to add a `replace=False` to `most_similiar` which is further passed down to `init_sims`. The user then has to change `replace` to `True` in order to be able to evaluate the newly calculated dimensions during training.
Further, this modification can be done without changing the current behavior.

menshikh-iv commented:
Abandoned, closing.

timbicker commented:
@menshikh-iv Hi sorry, I was busy in the last weeks. I will continue in the next days working on this PR, if it is still possible.

menshikh-iv commented:
@timbicker of course, feel free to re-open current PR and continue when you'll have time!

timbicker commented:
@menshikh-iv Hi, I just updated the PR but I can't reopen it. Can you reopen it for me?
Further, I added a `replace=False` to `most_similiar` which is further passed down to `init_sims`. The user then has to change `replace` to `True` in order to be able to evaluate the newly calculated vectors during training. This modification is imo necessary, because `_clear_post_train` is only called after all training is done, and therefore would not allow the user to use newly calculated keyedvectors in callbacks.

menshikh-iv commented:
@timbicker reopened, feel free to continue :+1:

gojomo commented:
Adding a `replace` to `most_similar()` seems unnecessarily complex, and inappropriate for a method that would otherwise have no side-effects on the model. 

Why not just clear the normed-vectors the moment any action that could change the raw vectors is initiated? (Pre-any-training-epoch, not post-any-training-epoch?) 

If some really-unique pattern of usage needs them cleared other times, that code can and should explicitly call `init_sims()` itself, rather than layering extra possibly-destructive side-effects into `most_similar()`.

timbicker commented:
Unfortunately `FastText` has problems with this implementation. 
```
Traceback (most recent call last):
  File "lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "gensim/gensim/models/base_any2vec.py", line 211, in _worker_loop
    tally, raw_tally = self._do_train_job(data_iterable, job_parameters, thread_private_mem)
  File "gensim/gensim/models/fasttext.py", line 610, in _do_train_job
    tally += train_batch_cbow(self, sentences, alpha, work, neu1)
  File "gensim/models/fasttext_inner.pyx", line 431, in gensim.models.fasttext_inner.train_batch_cbow
TypeError: 'NoneType' object is not subscriptable
```
I don't know how to fix this problem, because I am not familiar with fasttext...
A workaround would be, to check if `self` is of class `FastText` in `train()` and then call `clear_keyed_vector_internals` (former `_clear_post_train`) accordingly.
Can you help me out here? What are your thoughts?

timbicker commented:
I solved the problem with two functions `_clear_epoch_start` and `_clear_post_train`. Both functions do nothing in their original parent implementation and can be overridden accordingly by the child functions. This allows `Doc2Vec` and `Word2Vec` to implement their functionality independent of `FastText`

gojomo commented:
Personally still see this as a little over-wrought for the purpose:

* it clears *both* on epoch-start and training-finish
* (new-method) `_clear_epoch_start()` only ever calls `clear_sims()` in subclasses, and is empty in superclass, adding more twisty indirection

I'd prefer the more-simple (even if slightly more error-prone) approach of: (1) always clear at the end of `train()`; (2) warn via docs the (fewer, particular, plausibly-expert) users of the advanced callbacks functionality that *if* they're doing the kinds of operations that cache model results, they need to `clear_sims()` to get updated results.

timbicker commented:
@gojomo thanks for your reply. Unfortunately, I won't have time anymore to give this PR the attention it deserves. I would appreciate it if you could set the flag to `help wanted` or something you think appropriate.
Maybe @Naba7 wants to finish?

m4nuC commented:
I am using FastText and I've stumbled on this issue during training. 
I've tried setting the norms to `None` but that doesn't seem to help.
 I've also tried  `model.wv.clear_keyed_vector_internals` or `model.clear_keyed_vector_internals`  ? But neither method seem to exist.

Are there any other workaround that I should use ?

gojomo commented on gensim/models/fasttext.py:
```diff
@@ -745,7 +741,10 @@ def clear_sims(self):
+        self.wv.vectors_norm = None
+        self.wv.vectors_vocab_norm = None
+        self.wv.vectors_ngrams_norm = None
+        self.wv.buckets_word = None
```
This is fishy: nothing about training (as opposed to other vocab-discovery steps) would change the `buckets_word` mapping of words to ngram-locations. But also: (1) `init_sims()` doesn't rebuild this `buckets_word` as implied by the comment; (2) unless it's a big performance win in some usual case, I'm not sure why `buckets_word` needs to exist at all – I thought known-words already had presummed vectors (requiring no ngram lookups), while for any OOV word, it's not very expensive to find the ngrams/ngram-bucket-indexes each time needed; (3) relatedly, it looks like the only use of `vectors_ngrams_norm` is nonsensical, so that property should be discarded entirely - but I've made a separate issue, #2667, for that.

mpenkov commented:
I think this is superseeded by #2698. Closing.
